pkg/services/secrets/kvstore/plugin_mig.go

```diff
@@ -0,0 +1,75 @@
+package kvstore
+
+import (
+	"context"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/services/secrets"
+	"github.com/grafana/grafana/pkg/services/sqlstore"
+	"github.com/grafana/grafana/pkg/setting"
+)
+
+// PluginSecretMigrationService This migrator will handle migration of datasource secrets (aka Unified secrets)
+// into the plugin secrets configured
+type PluginSecretMigrationService struct {
+	secretsStore   SecretsKVStore
+	cfg            *setting.Cfg
+	logger         log.Logger
+	sqlStore       sqlstore.Store
+	secretsService secrets.Service
+	remoteCheck    UseRemoteSecretsPluginCheck
+}
+
+func ProvidePluginSecretMigrationService(
+	secretsStore SecretsKVStore,
+	cfg *setting.Cfg,
+	sqlStore sqlstore.Store,
+	secretsService secrets.Service,
+	remoteCheck UseRemoteSecretsPluginCheck,
+) *PluginSecretMigrationService {
+	return &PluginSecretMigrationService{
+		secretsStore:   secretsStore,
+		cfg:            cfg,
+		logger:         log.New("sec-plugin-mig"),
+		sqlStore:       sqlStore,
+		secretsService: secretsService,
+		remoteCheck:    remoteCheck,
+	}
+}
+
+func (s *PluginSecretMigrationService) Migrate(ctx context.Context) error {
+	// Check if we should migrate to plugin - default false
+	if s.cfg.SectionWithEnvOverrides("secrets").Key("migrate_to_plugin").MustBool(false) &&
+		s.remoteCheck.ShouldUseRemoteSecretsPlugin() {
+		// we need to instantiate the secretsKVStore as this is not on wire, and in this scenario,
+		// the secrets store would be the plugin.
+		secretsSql := &secretsKVStoreSQL{
+			sqlStore:       s.sqlStore,
+			secretsService: s.secretsService,
+			log:            s.logger,
+			decryptionCache: decryptionCache{
+				cache: make(map[int64]cachedDecrypted),
+			},
+		}
+
+		allSec, err := secretsSql.GetAll(ctx)
+		if err != nil {
+			return nil
+		}
+		// We just set it again as the current secret store should be the plugin secret
+		for _, sec := range allSec {
+			err = s.secretsStore.Set(ctx, *sec.OrgId, *sec.Namespace, *sec.Type, sec.Value)
+			if err != nil {
+				return err
+			}
+		}
+		// as no err was returned, when we delete all the secrets from the sql store
+		for _, sec := range allSec {
```

Comment:
Do you think it might be worth creating a DelAll function here instead of looping again?

Reply:
That is not a bad idea, it would ensure that the process is not broken in the middle, and we are left with half of the secrets deleted. I will add that for this PR as it seems important

Reply:
Yes, I agree. We need a bulk get and a bulk delete to make the migration more straight forward.

Such methods might be retroactively implemented on the Data Source migration in a different PR after this one is merged.

Reply:
Added this issue to work on it later https://github.com/grafana/grafana-partnerships-team/issues/275